### DISCUSSION TITLE
osu!mania SR discrepancy when line orders changed

### DISCUSSION THREAD

**Discussion by aryoadhi** [ppy/osu#22745] [Q&A] [answered]
Taken from this help forum thread: https://osu.ppy.sh/community/forums/topics/1727438

It seems that osu!mania SR suffered from discrepancies whenever the contents below `[HitObjects]` are rearranged or shuffled. Issue persist on both the web and the game (stable, and lazer, quoted from recent replies https://github.com/ppy/osu/discussions/22745#discussioncomment-5129676).

Below is the example of the beatmap where the bug occurs: 
* https://osu.ppy.sh/beatmapsets/1945296

There are multiple ways that this happened, but for me this method is almost a guaranteed hit and not done by modifying `.osu` file deliberately using text editor. The steps below assumes in-game editor of osu!stable.
1. Create two copies of an osu!mania beatmap. You can rename the difficulty as anything, but for this example rename one as `Control` and `Experiment`.
2. Leave the `Control` as-is.
3. For `Experiment`, do a full cut-and-paste of the map (Ctrl+A, Ctrl+X, Ctrl+V) on the very first note of the map. This ensures the map to be structurally same and doesn't have any offset differences. Make sure to save the map after doing so.
4. Compare the SR between `Control` and `Experiment`.

Mostly, the SR discrepancy is negligible, but I managed to (accidentally) bugged the SR system that it has ±0.3 SR change.

![Example](https://cdn.discordapp.com/attachments/1060170699500691466/1078486366645657790/image.png)

As a further confirmation, I wrote a Python code to verify if there are no unique lines between `Control` and `Experiment` map, which you can see on my gist: https://gist.github.com/aryoadhi/647fca999423de226feced3c24155a0d

We need to fix this issue so that the SR integrity of osu!mania's mapsets become consistent. Worst possible case of this is someone deliberately exploiting this bug to achieve higher-than-intended SR on a map, creating a so-called "PP farm" map.

**Comment by bdach** [comment]
Please check if this affects lazer first. I think after changes like https://github.com/ppy/osu/pull/19120 there's a good chance it will not.

The fix will not be backported to stable, I'm pretty sure. Stable's diffcalc/ppcalc code is EOL and does not receive further changes.

**Reply by aryoadhi** [reply]
It also affect osu!lazer.

![Screenshot_2023-02-28-05-00-28-48_25a1a32208bbcdc1d450b7aa854bb161](https://user-images.githubusercontent.com/5029654/221695759-ee4702a9-2c09-48c9-952e-2eb71b1ae8ca.jpg)
![Screenshot_2023-02-28-05-00-33-34_25a1a32208bbcdc1d450b7aa854bb161](https://user-images.githubusercontent.com/5029654/221695778-69fb1651-9567-475f-b800-9e63d193723f.jpg)

**Reply by bdach** [reply]
Can you attach the beatmaps you were testing with?

**Reply by aryoadhi** [reply]
I only tested one, and so far this is the one with major discrepancy.

https://osu.ppy.sh/beatmapsets/1945296

It seems that more complex the pattern is, then the fluctuation range expands further. This probably works better on hybrid maps, as the difference on rice maps are very negligible.

**Reply by aryoadhi** [reply]
I put this into an even more extreme test by deliberately shuffling the note line orders. It does fluctuate the SR by a certain degree. The map itself is still structurally same (i.e played exactly the same).

![image](https://user-images.githubusercontent.com/5029654/221836231-eea423f0-66c4-463b-becd-d6687db95916.png)

In addition, with my proposed in-game method to skew the SR, the stable editor's copy-paste algorithm seems caused the map's SR to alternate between two values, which is the original and the skewed version. For this case, the difficulty `Insane (NSV)` and `Hyper Insane` will have their SR change to each other when processed, and revert to the original when done again.

**Comment by molneya** [comment, marked as answer]
This is caused by https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs#L58-L59. Depending on the note processing order, some long notes won't update the `holdFactor` when expected. The fix for this should either be to enforce longer long notes to be processed first, or change the condition to ignore long notes that start at the same time. The exact solution (and how to balance the changes) are still to be worked out by the mania pp committee.

**Reply by aryoadhi** [reply]
Can we implement I/O solution that enforces the .osu file to be 'fixed' on save instead of fixing the ruleset? For example, if we 'save' the osu!mania beatmap, the editor does sanity check so the lines are ordered by this, in priority:
1. Note position, in miliseconds
2. Note position, in x,y (column, leftmost first)

In case for notes with special keys at point 2, the S key is placed topmost before other notes.

**Reply by molneya** [reply]
It shouldn't matter what order concurrent notes are processed in. If a certain column order was enforced, it would just result in bias towards certain columns (such as mirroring a map might change star rating), not solving anything. It must be fixed at the pp/sr level.

**Reply by aryoadhi** [reply]
It's a bit off-topic, but the idea of the beatmap standardization as I proposed is to tidy up the beatmap structure. The sanity check supposedly not interfere with the SR as you mentioned before, since the formula supposedly discern the order of notes.

On another note, where does the behavior of SR osciliating come from? Does osu! handle mania copy-paste in a certain way? (see https://github.com/ppy/osu/discussions/22745#discussioncomment-5145100)

**Reply by osuAntalf** [reply]
> This is caused by https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs#L58-L59. Depending on the note processing order, some long notes won't update the `holdFactor` when expected. The fix for this should either be to enforce longer long notes to be processed first, or change the condition to ignore long notes that start at the same time. The exact solution (and how to balance the changes) are still to be worked out by the mania pp committee.

I have also tested this with maps that have 0 long notes and I also got a drop in difficulty but not an increase. You can test this on https://osu.ppy.sh/beatmapsets/224173#mania/523391 and get this result https://osu.ppy.sh/ss/18464809/712b

Would this also mean that the same is happening for single notes or normal notes as well?

**Reply by aryoadhi** [reply]
I can't reproduce this in lazer. The issue must be beatmaps with long notes.

![osu_2023-03-01_01-20-58](https://user-images.githubusercontent.com/5029654/221944494-94a0bb5e-c7cd-417c-8495-23269d102f54.jpg)

I wrote a program to automate the line shuffling process and all the SR is exact same of 6.18.

**Reply by aryoadhi** [reply]
By glitching the same map I mentioned on the first post (https://osu.ppy.sh/beatmapsets/1945296), but instead with deliberate line shuffling, I can confirm confidently that it is related to the LNs and how does the map is structured. By blanking the `BeatmapID` and `BeatmapSetID`, I'm sure that the bugged beatmap doesn't derive its SR from the existing uploaded one.

![osu_2023-03-01_06-41-28](https://user-images.githubusercontent.com/5029654/222009130-b3d1a9a2-fba3-404e-9ba1-1fee6799cf4c.jpg)
![osu_2023-03-01_06-41-42](https://user-images.githubusercontent.com/5029654/222009143-386d74ce-7a2b-4e94-8bf2-f71dd6350655.jpg)

The concern about potential abuse is probably not as much since we confirmed that maps with 0 LN won't be affected, and existing ranked maps won't submit score unless it's a verbatim copy of the server's version. But still, it's a proof-of-concept that the SR system can be exploited.

**Reply by peppy** [reply]
https://github.com/ppy/osu/issues/22756

**Comment by osuAntalf** [comment]
This issue was brought up to me by the user and I have done some testing myself and figured out something. When maps contain heavy usage of long notes sometimes the difficulty will drop. Take into example this map (will do it in links as to not clutter this comment):

https://osu.ppy.sh/ss/18464739/758b

The initial SR is 5.75 but when I recreated the glitch the difficulty dropped significantly to 5.52 (0.23 from the original difficulty):

https://osu.ppy.sh/ss/18464741/5f96

My main concern was that if this could be infinitely repeated this could cause an exponential increase in the difficulty by being done over and over again. For the sake of this test I decided to keep doing it on the previously copied difficulty now at 5.52 and this was the result:

https://osu.ppy.sh/ss/18464746/35df

The difficulty has jumped from 5.52 to 6.15. 0.63 from the copied difficulty and 0.40 from the original difficulty. This is mostly present when maps have a high density of long notes. My main concern for this issue is that if this can be replicated consistently through out _any_ map this can cause an inflation in SR and in turn a PP inflation as well. As a beatmap nominator this is very concerning since a map can look more difficult than the SR suggest so it is very easy to get fooled by this if someone does not have a keen eye.

The process to inflate the map to this level required 3 steps, some will stop at 2 and even 1 copy. If anyone would like me to attach the map that I was testing this on feel free to reply.

**Reply by molneya** [reply]
Please attach the beatmap. If the glitch can actually cause an exponential increase in difficulty that is something that needs to be looked into.

**Reply by osuAntalf** [reply]
Something else worth mentioning would be that I also tested this with ranked maps. You could buff the SR by doing this, update the map and it would stay locally by the amount it was buffed. Could this also create an increment of PP in already-ranked maps? Since the difficulty did stay the same even after changing the name and overwriting the ranked one so a play could be made with the new SR and affect the amount of PP one could gain?

**Reply by aryoadhi** [reply]
It shouldn't be possible. Even locally saving the file while nothing in the map changed at all (e.g only changed editor settings such as zoom or snapping) will ask me to update the map before playing it for ranked.

In other words, the `.osu` content must be exactly the same as the ones in the server.

**Reply by osuAntalf** [reply]
Here is the [beatmap](https://www.dropbox.com/s/ki78i7mek2ijv7s/Shiraishi%20-%20True%20Millennium%20Queen.osz?dl=0). Please be advised to not update the map since this will delete it completely. I've also left the replicated difficulties in case anyone would like to take a look at them, if no, delete and repeat the same process.

If anyone manages to get the same result as I did, a reply to this will be greatly appreciated.

**Reply by molneya** [reply]
The beatmap has the same issue as https://github.com/ppy/osu/discussions/22745#discussioncomment-5148400. It is fixed by the solutions mentioned there. Whatever you are seeing with "star rating exponentially increasing" may be a some kind of stable bug. It definitely won't cause beatmaps to be ranked with star ratings way higher than expected, for example.

Also, If you want to point out bugs in the difficulty algorithm, please do so on the lazer client, since stable's local star rating isn't being updated any more and is outdated.

**Reply by aryoadhi** [reply]
Would uploading the beatmap to listing also count as 'testing in lazer'? The formula for web and lazer SR calculation seems to be the same.

**Reply by molneya** [reply]
> The formula for web and lazer SR calculation seems to be the same.

If there is an update to lazer SR that hasn't been released to web, then they might not be the same. Just test on the latest version of lazer as it will always be up to date with the latest algorithm.

**Comment by peppy** [comment]
I can't follow the threaded conversations in here, but as far as I can tell this is a bug with difficulty calculation, and potentially a recent change if this is not being noticed on live systems? If so, it should be escalated to an issue.

**Reply by aryoadhi** [reply]
Yes. I think it's already raised as an issue. https://github.com/ppy/osu/issues/22756

How does this impact osu!mania PP calculation if the fix is applied in the future?
